### PR TITLE
Add upload file with usages

### PR DESCRIPTION
**Pull request by szok**

### CODE REVIEW

File: src/ralph_scrooge/media/scrooge/app/allocationclient/controller.js [hunk 1]
```diff
@@ -25,6 +25,44 @@ scrooge.controller('allocationClientCtrl', ['$scope', '$routeParams', '$http', '
...
+
+        var uploadUrl = urlClientChunks.join('/');
+        var file = $scope.myFile;
+        var fd = new FormData();
+        fd.append('file', file);
+        $http.post(uploadUrl, fd, {
+            transformRequest: angular.identity,
+            headers: {'Content-Type': undefined}
+        })
+        .success(function(data) {
```
Comment (mkurek) [inline src/ralph_scrooge/media/scrooge/app/allocationclient/controller.js]: move it to the line above

File: src/ralph_scrooge/media/scrooge/app/allocationclient/controller.js [hunk 2]
```diff
@@ -25,6 +25,44 @@ scrooge.controller('allocationClientCtrl', ['$scope', '$routeParams', '$http', '
...
+            if (data['status'] == true) {
+                $('#upload_modal').modal('hide');
+            } else {
+                alert(data['error']);  // TODO: Alert?
+            }
+        })
+        .error(function(){
```
Comment (mkurek) [inline src/ralph_scrooge/media/scrooge/app/allocationclient/controller.js]: ditto

File: src/ralph_scrooge/media/scrooge/partials/taballocationclientdivision.html [hunk 3]
```diff
@@ -1,7 +1,9 @@
+
+<br />
 <table class="allocation-admin-table">
   <thead>
     <tr>
-      <th>Charged service</th>
```
Comment (mkurek) [inline src/ralph_scrooge/media/scrooge/partials/taballocationclientdivision.html:4]: why?

File: src/ralph_scrooge/utils/csv_utils.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
+import csv
```
Comment (mkurek) [inline src/ralph_scrooge/utils/csv_utils.py]: unify this with https://github.com/allegro/ralph_pricing/blob/develop/src/ralph_scrooge/csvutil.py

File: src/ralph_scrooge/rest/allocationclient.py [hunk 5]
```diff
@@ -206,7 +207,39 @@ def get(self, request, year, month, service, env, *args, **kwargs):
 
     @commit_on_success()
     def post(self, request, year, month, service, env, *args, **kwargs):
-        post_data = json.loads(request.raw_post_data)
+        if request.FILES and kwargs.get('allocate_type') == 'servicedivision':
+            file_results = []
+            data_results = parse_csv(request.FILES['file'])
+            for row in data_results:
+                query = {}
+                if row.get('service_uid'):
```
Comment (mkurek) [inline src/ralph_scrooge/rest/allocationclient.py]: please add option for passing service_env_id too (for some backward-compatibility)

**Review by mkurek — CHANGES_REQUESTED** [review]

File: src/ralph_scrooge/media/scrooge/partials/taballocationclientdivision.html (hunk 3 above)
Comment (szok) [reply src/ralph_scrooge/media/scrooge/partials/taballocationclientdivision.html:4]: My mistake :(

File: src/ralph_scrooge/rest/allocationadmin.py [hunk 6]
```diff
@@ -246,19 +288,22 @@ def _save_extra_costs(self, start, end, post_data):
...
+                    else:
+                        try:
+                            service_environment = ServiceEnvironment.objects.get(  # noqa
+                                service_id=ec_row['service'],
+                                environment_id=ec_row['env']
+                            )
+                        except ServiceEnvironment.DoesNotExist:
+                            raise ServiceEnvironmentDoesNotExistError(
+                                'No service environment with service id {0}'
+                                ' and environment id {1}'.format(
```
Comment (xor-xor) [inline src/ralph_scrooge/rest/allocationadmin.py]: What about `Service environment does not exist for service with ID {0} and environment with ID {1}`..?

File: src/ralph_scrooge/rest/allocationadmin.py [hunk 7]
```diff
@@ -52,6 +52,48 @@ class ServiceEnvironmentDoesNotExistError(Exception):
     pass
 
 
+def get_allocationadmin_from_file(file):
```
Comment (xor-xor) [inline src/ralph_scrooge/rest/allocationadmin.py]: What about adding a note somewhere re: how structure of this file should look like..? I mean, I know it's CSV, but what to you expect in its columns?

File: src/ralph_scrooge/rest/allocationclient.py [hunk 8]
```diff
@@ -31,13 +29,51 @@
...
 )
+from ralph_scrooge.csvutil import parse_csv
 
 
 class CannotDetermineValidServiceUsageTypeError(APIException):
     status_code = 400
     default_detail = 'Cannot determine valid (single!) service usage type'
 
 
+def get_allocation_from_file(file):
```
Comment (xor-xor) [inline src/ralph_scrooge/rest/allocationclient.py]: Same remark as in `get_allocationadmin_from_file` (i.e. description of records).

File: src/ralph_scrooge/rest/allocationadmin.py [hunk 9]
```diff
@@ -52,6 +52,69 @@ class ServiceEnvironmentDoesNotExistError(Exception):
...
+                },
+                ...
+            ]
+    """
+    file_results = []
+    data_results = parse_csv(file)
+    errors = []
+    for row in data_results:
+        query = {}
+        if row.get('service_env_id'):
```
Comment (mkurek) [inline src/ralph_scrooge/rest/allocationadmin.py]: we should first check (validate) if all required keys (cost, forecast_cost (it shouldn't be required imo - we could assume 0 here as a default), service_env_id or service_uid or service_name + environment) are set. We could also check headers of csv;

File: src/ralph_scrooge/rest/allocationadmin.py [hunk 10]
```diff
@@ -52,6 +52,69 @@ class ServiceEnvironmentDoesNotExistError(Exception):
...
+            errors.append(error)
+            service_env = ''
+
+        row_data = {
+            'service': service_env,
+            'env': service_env,
+            'forecast_cost': row['forecast_cost'],
+            'cost': row['cost']
+        }
+        if len(row.get('id', '')) > 1:
```
Comment (mkurek) [inline src/ralph_scrooge/rest/allocationadmin.py]: why do we need this?

File: src/ralph_scrooge/rest/allocationclient.py [hunk 11]
```diff
@@ -31,13 +29,72 @@
...
+                row.get('service_uid', row.get('service_name')),
+                row['environment']
+            )
+            errors.append(error)
+            service_env = ''
+
+        file_results.append({
+            'service': service_env,
+            'env': service_env,
+            'value': row['value']
```
Comment (mkurek) [inline src/ralph_scrooge/rest/allocationclient.py]: this differs from `get_allocationadmin_from_file` only by `value` (cost/forecast_cost in `get_allocationadmin_from_file`) - could we unify this?

File: src/ralph_scrooge/rest/allocationclient.py [hunk 12]
```diff
@@ -206,7 +263,17 @@ def get(self, request, year, month, service, env, *args, **kwargs):
...
+        if request.FILES and kwargs.get('allocate_type') == 'servicedivision':
+            rows, errors = get_allocation_from_file(request.FILES['file'])
+            if errors:
+                return Response({'status': False, 'errors': errors})
+
+            post_data = {
+                'rows': rows
+            }
+        else:
+            post_data = request.DATA
```
Comment (mkurek) [inline src/ralph_scrooge/rest/allocationclient.py:310]: shouldn't we `json.loads` this? what's the difference between `request.raw_post_data` and `request.DATA`?

**Review by mkurek — CHANGES_REQUESTED** [review]

File: src/ralph_scrooge/rest/allocationadmin.py (hunk 10 above)
Comment (szok) [reply src/ralph_scrooge/rest/allocationadmin.py]: Yes, because I don't add to result if 'id' is not exist.

File: src/ralph_scrooge/csvutil.py [hunk 13]
```diff
@@ -108,3 +108,34 @@ def make_csv_response(data=[], filename='export.csv', encoding='cp1250'):
...
+    result = []
+    rows = list(reader)
+    headers = rows[0]
+    for row in rows[1:]:
+        line = {}
+        for i, item in enumerate(headers):
+            line[item] = row[i]
+        result.append(line)
+
+    return (result, headers)
```
Comment (mkurek) [inline src/ralph_scrooge/csvutil.py]: invert it - return headers first

**Review by mkurek — CHANGES_REQUESTED** [review]

File: src/ralph_scrooge/csvutil.py (hunk 13 above)
Comment (szok) [reply src/ralph_scrooge/csvutil.py]: done

**Review by mkurek — APPROVED** [review]

**Pull request merged**